edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/MqttBrokerConnectorTest.cs

```diff
@@ -287,17 +287,31 @@ public async Task WhenReconnectsThenResubscribes()
 
             broker.Subscriptions.Clear();
 
-            var milestone = new SemaphoreSlim(0, 1);
-            broker.OnConnect = () => milestone.Release();
+            var milestoneConnected = new SemaphoreSlim(0, 1);
+            broker.OnConnect = () => milestoneConnected.Release();
+
+            var milestoneSubscribed = new SemaphoreSlim(0, 1);
+            var subscriptionCount = consumers.SelectMany(c => c.Subscriptions).Count();
+            broker.OnSubscribe = () =>
+            {
+                if (Interlocked.Decrement(ref subscriptionCount) == 0)
+                {
+                    milestoneSubscribed.Release();
+                }
+            };
+
             broker.DropActiveClient();
 
-            Assert.True(await milestone.WaitAsync(TimeSpan.FromSeconds(5)));
+            Assert.True(await milestoneConnected.WaitAsync(TimeSpan.FromSeconds(5)));
+            Assert.True(await milestoneSubscribed.WaitAsync(TimeSpan.FromSeconds(5)));
 
-            var expected = consumers.SelectMany(c => c.Subscriptions).OrderBy(s => s);
-            Assert.Equal(expected, broker.Subscriptions.OrderBy(s => s));
+            var expected = consumers.SelectMany(c => c.Subscriptions).OrderBy(s => s).ToArray();
+            var actual = broker.Subscriptions.OrderBy(s => s).ToArray();
+
+            Assert.Equal(expected, actual);
```

Comment:
I'm not sure if I follow the logic of this test. Specifically, where is the guarantee that you'll wait until client re-subscribes before doing assertion?

Reply:
the logic for waiting for re-subscribe is in the broker.OnSubscribe(). First it counts how many subscription it should wait for, then every subscription decreases the counter and at the end the semaphore releases, that is what a bit lower the code is waiting for.